xwiki-platform-core/xwiki-platform-refactoring/xwiki-platform-refactoring-default/src/main/java/org/xwiki/refactoring/job/XClassDeletingListener.java

```diff
@@ -0,0 +1,172 @@
+            logger
+                .warn("XClass deleting listener will not check the document in non-interactive mode.");
+            return;
+        }
+
+        // Check if some pages contain used XClass
+        Map<EntityReference, EntitySelection> concernedEntities = (Map<EntityReference, EntitySelection>) data;
+        XClassBreakingQuestion question = new XClassBreakingQuestion(concernedEntities);
+        for (EntitySelection entitySelection : concernedEntities.values()) {
+            if (entitySelection.getEntityReference() instanceof DocumentReference) {
+                checkIfDeleteIsAllowed(entitySelection, question);
+            }
+        }
+
+        // Ask a confirmation to the user if some pages contain used XClass
+        if (!question.getImpactedObjects().isEmpty()) {
+
+            try {
+                if (this.documentAccessBridge.isAdvancedUser(job.getRequest().getProperty("user.reference"))) {
+                    // Conservative choice: we let the user enable the pages to delete.
+                    question.unselectAll();
+                    // The user can modify the question so it could disable some EntitySelection.
+                    // We add a timeout because when a refactoring job is running, it prevents others to run.
+                    // 5 minutes is probably enough for the user to decide if the process should go on.
+                    boolean ack = job.getStatus().ask(question, 5, TimeUnit.MINUTES);
+                    if (!ack) {
+                        // Without any confirmation, we must cancel the operation.
+                        String message = "The question has been asked, however no answer has been received.";
+                        this.logger.warn(message);
+                        cancelableEvent.cancel(message);
+                    }
+                } else {
+                    question.setRefactoringForbidden(true);
+                    // we don't want the user to answer the question,
+                    // but we want to display that his action is forbidden.
+                    boolean ack = job.getStatus().ask(question, 1, TimeUnit.MINUTES);
+                    if (!ack) {
+                        String message = "The question has been canceled because this refactoring is forbidden.";
+                        cancelableEvent.cancel(message);
+                    }
+                }
+            } catch (InterruptedException e) {
+                this.logger.warn("Confirm question has been interrupted.");
+                cancelableEvent.cancel("Question has been interrupted.");
+            }
+        }
+    }
+
+    private void checkIfDeleteIsAllowed(EntitySelection entitySelection, XClassBreakingQuestion question)
+    {
+        DocumentReference documentReference = (DocumentReference) entitySelection.getEntityReference();
+        String query = "select distinct doc.fullName from XWikiDocument doc, BaseObject obj where "
```

Comment:
select distinct obj.name from BaseObject obj where obj.className=:className

would be faster.

Reply:
It could technically but that would mean the database is not in a great shape. At worst if there is ghosts it just mean you will get a warning when deleting a xclass which is kind of not really used.

Reply:
>  At worst if there is ghosts it just mean you will get a warning 

IMO the worst would be to get stucked when trying to delete an unused XClass because of a broken UI because of some remaining ghosts...

Reply:
Why broken ? The UI will just tell you that the class is used and you conform you still want to delete it. But again it would mean someone purposely deleted an entry of the the document table without going trough the API. IMO the very small risk of useless warning worth the potential important performance boost in big databases (plus this optimization is often used).